client/messaging/message-list.tsx

```diff
@@ -41,34 +35,45 @@ const Messages = styled.div`
 `
 
 interface PureMessageListProps {
-  messages: List<ChatMessage>
+  messages: List<Message>
+  renderMessage?: (msg: Message) => ReactNode
 }
 
-function messageToElem(msg: ChatMessage) {
+function renderCommonMessage(msg: Message) {
   switch (msg.type) {
-    case ChatMessageType.JoinChannel:
-      return <JoinChannelMessage key={msg.id} record={msg} />
-    case ChatMessageType.LeaveChannel:
-      return <LeaveChannelMessage key={msg.id} record={msg} />
-    case ChatMessageType.TextMessage:
+    case CommonMessageType.TextMessage:
       return <TextMessageDisplay key={msg.id} user={msg.from} time={msg.time} text={msg.text} />
-    case ChatMessageType.NewChannelOwner:
-      return <NewChannelOwnerMessage key={msg.id} record={msg} />
-    case ChatMessageType.SelfJoinChannel:
-      return <SelfJoinChannelMessage key={msg.id} record={msg} />
     default:
       return null
   }
 }
 
+const handleUnknown = (msg: Message) => {
+  throw new Error('Trying to render unknown message type: ' + msg.type)
```

Comment:
I'm not sure it really makes sense to do this, we didn't enforce that all message types got rendered before. This also doesn't really offer any protection from anything, as it only ever gets called if `renderMessage` is not passed. Converting a situation that doesn't show a message into one that breaks the applications seems like a not great idea to me. If we could check that all cases were handled at compile-time, fine, but we can't really do that with how these messages are split up now.

So, tl;dr: just make this an empty function that returns null